RFC-0020-Lightweight-Dispatch.md

```diff
@@ -0,0 +1,291 @@
+    * With codegen unboxing + static dispatch, we hope that we can reach a much smaller percentage of cycle count for the op registration step.
+
+
+
+#### Avoid schema parsing at runtime
+
+As mentioned in step 1, we are registering an operator into the registry along with a schema string. We realized at the library initialization time we need to spend a lot of resources on schema parsing, according to the profiling results based on our prototype. We also noticed that the required information to instantiate a schema object are all available at codegen time, we can pass these data to the registry directly so that we can save time at runtime. For example:
+
+
+```
+CodegenUnboxing.cpp
+===================
+RegisterOperators reg({
+   OperatorGenerator(
+       "aten::get_device", // name
+       "", // overload_name
+       arguments, // a vector of arguments
+       returns, // a vector of returns
+       [](Stack & stack) {
+         RECORD_FUNCTION("get_device", std::vector<c10::IValue>());
+         at::unboxing::get_device(stack);
+       },
+       aliasAnalysisFromSchema()
+   ),
+...
+})
+```
+
+
+This way we can directly instantiate `FunctionSchema` objects without parsing at runtime. Of course we need to change APIs in `operator.h` to make this happen.
+
+Q: Can we completely get rid of `FunctionSchema` and only register name/overload_name?
+
+A: No, because we should have feature parity to the current system and backward compatibility for mobile models is a feature we need to support for the lightweight dispatch system. Currently we rely on the number of arguments to let the new runtime be able to run the old model.
+
+
+#### Support tracing based selective build
+
+* In [gen.py](https://github.com/pytorch/pytorch/blob/master/tools/codegen/gen.py) the files we generate will go through the selector similar to what we are doing to `RegisterSchema.cpp` right now.
+* We need to make sure the binary size is on-par with or even better than existing tracing based selective build.
+
+## Risks
+
+There are 3 risks:
+
+
+
+1. Performance gain of using JIT op registry is insignificant or even worse than dispatcher.
+    1. De-risked: from the prototype running on a target platform it is proved to save latency on initial load.
+2. Binary size regression. Need to make sure selective build works.
+3. Mobile use case requires features only available on dispatcher.
+    1. E.g., boxed fallback mechanism for [conj](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/MathBitsFallback.h#L78) operator.
+
+## Testing & Tooling Plan
+
+Expand existing tests on lite interpreter to cover codegen logic. Let the cpp test target depending on codegen unboxing library, test if a module forward result from JIT execution equals to the lite interpreter execution. Since JIT execution goes through metaprogramming unboxing and lite interpreter execution goes through codegen unboxing, we can make sure the correctness of codegen unboxing. Example:
```

Comment:
```suggestion
Expand existing tests on the mobile interpreter to cover codegen logic. Let the cpp test target depending on codegen unboxing library, test if a module forward result from JIT execution equals to the mobile interpreter execution. Since JIT execution goes through metaprogramming unboxing and mobile interpreter execution goes through codegen unboxing, we can make sure the correctness of codegen unboxing. Example:
```